File: pt5/1130.minimum-cost-tree-from-leaf-values.java
import java.util.*;

class Solution {
    public int mctFromLeafValues(int[] arr) {
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(Integer.MAX_VALUE);
        int cost = 0;

        for (int value : arr) {
            while (stack.peek() <= value) {
                int middle = stack.pop();
                cost += middle * Math.min(stack.peek(), value);
            }
            stack.push(value);
        }

        while (stack.size() > 2) {
            cost += stack.pop() * stack.peek();
        }

        return cost;
    }
}

File: pt5/1130.minimum-cost-tree-from-leaf-values.py
from __future__ import annotations

from typing import List


class Solution:
    def mctFromLeafValues(self, arr: List[int]) -> int:
        stack = [float("inf")]
        cost = 0

        for value in arr:
            while stack[-1] <= value:
                middle = stack.pop()
                cost += middle * min(stack[-1], value)
            stack.append(value)

        while len(stack) > 2:
            cost += stack.pop() * stack[-1]

        return cost

